cpp/src/arrow/compute/kernels/hash_aggregate_test.cc

```diff
@@ -2460,6 +2460,476 @@ TEST(GroupBy, Distinct) {
+                                      }),
+                                      R"([
+    [1.0,  null, true,  1],
+    [0.0,  null, false, 2],
+    [null, null, false, 3],
+    [4.0,  null, null,  null]
+  ])"),
+                        aggregated_and_grouped,
+                        /*verbose=*/true);
+    }
+  }
+}
+
+TEST(GroupBy, OneTypes) {
+  std::vector<std::shared_ptr<DataType>> types;
+  types.insert(types.end(), NumericTypes().begin(), NumericTypes().end());
+  types.insert(types.end(), TemporalTypes().begin(), TemporalTypes().end());
+  types.push_back(month_interval());
+
+  const std::vector<std::string> default_table = {R"([
+    [1,    1],
+    [null, 1]
+])",
+                                                  R"([
+    [0,    2],
+    [null, 3],
+    [3,    4],
+    [5,    4],
+    [4,    null],
+    [3,    1],
+    [0,    2]
+])",
+                                                  R"([
+    [0,    2],
+    [1,    null],
+    [null, 3]
+])"};
+
+  const std::vector<std::string> date64_table = {R"([
+    [86400000,  1],
+    [null,      1]
+])",
+                                                 R"([
+    [0,         2],
+    [null,      3],
+    [259200000, 4],
+    [432000000, 4],
+    [345600000, null],
+    [259200000, 1],
+    [0,         2]
+])",
+                                                 R"([
+    [0,         2],
+    [86400000,  null],
+    [null,      3]
+])"};
+
+  const std::string default_expected =
+      R"([
+    [1,    1],
+    [0,    2],
+    [null, 3],
+    [3,    4],
+    [4,    null]
+    ])";
+
+  const std::string date64_expected =
+      R"([
+    [86400000,  1],
+    [0,         2],
+    [null,      3],
+    [259200000, 4],
+    [345600000, null]
+    ])";
+
+  for (const auto& ty : types) {
+    SCOPED_TRACE(ty->ToString());
+    auto in_schema = schema({field("argument0", ty), field("key", int64())});
+    auto table =
+        TableFromJSON(in_schema, (ty->name() == "date64") ? date64_table : default_table);
+
+    ASSERT_OK_AND_ASSIGN(
+        Datum aggregated_and_grouped,
+        GroupByTest({table->GetColumnByName("argument0")},
+                    {table->GetColumnByName("key")}, {{"hash_one", nullptr}},
+                    /*use_threads=*/false, /*use_exec_plan=*/true));
+    ValidateOutput(aggregated_and_grouped);
+    SortBy({"key_0"}, &aggregated_and_grouped);
+
+    AssertDatumsEqual(
+        ArrayFromJSON(struct_({
+                          field("hash_one", ty),
+                          field("key_0", int64()),
+                      }),
+                      (ty->name() == "date64") ? date64_expected : default_expected),
+        aggregated_and_grouped,
+        /*verbose=*/true);
+  }
+}
+
+TEST(GroupBy, OneDecimal) {
+  auto in_schema = schema({
+      field("argument0", decimal128(3, 2)),
+      field("argument1", decimal256(3, 2)),
+      field("key", int64()),
+  });
+  for (bool use_exec_plan : {false, true}) {
+    for (bool use_threads : {/*true, */ false}) {
```

Comment:
The `true` option for `use_threads` is greyed out in all these test because, well, they run parallelly, and the way `hash_one` is currently implemented, when it encounters a value (null or otherwise) for a group for the first time, it stores that value and never updates it.
For example, in the table below, 
- If we go serially, `["4.01", "4.01",   null]` is the first row encountered which has `null` as its key. But,
- If we go parallelly, `["0.75",  "0.75",  null]` is the first row encountered which has `null` as its key.

And while this is totally expected, this is a bit tricky to test for in a single go. So should we test explicitly for both `use_threads` = `true` and `false`?

Reply:
Hmm, we can't assume any particular value and we shouldn't assume anything about the order that the kernel gets batches whether it's serial or parallel. That makes the test hard to write, though.

We could do something like this:
```cpp
const auto& struct_arr = aggregated_and_grouped.array_as<StructArray>();
// Check the key column
AssertDatumsEqual(ArrayFromJSON(int64(), "[1, 2, 3, 4, null]"), struct_arr->field(2)));
// Check values individually
const auto& col0 = struct_arr->field(0);
ASSERT_OK_AND_ASSIGN(const auto col0_0, col0->GetScalar(0));
EXPECT_THAT(col0_0->Equals(*ScalarFromJSON(...)) || col0_0->Equals(*ScalarFromJSON()) ...);
```

this would get tedious fast though so I would honestly reduce the number of groups in the first place

Reply:
Or you can try to make AnyOf work with EXPECT_THAT: https://github.com/google/googletest/blob/main/docs/reference/matchers.md#composite-matchers